File: Easy/331_adding_calculator.rb
# dailyprogrammer 331: adding calculator. Requires rubinius 3.84
class AddingCalculator
  def initialize
    input_loop
  end

  private

  NEG_VAL = Fixnum::MIN + 4_611_686_018_427_387_902

  def input_loop
    loop do
      print '> '
      @input = gets.chomp
      check_input
    end
  end

  def check_input
    @arr = @input.split(' ')
    @a = @arr[0].to_i
    @b = @arr[2].to_i
    @operator = @arr[1]
    evaluate
  end

  def evaluate
    case @operator
    when '*' then puts multiply(@a, @b)
    when '/' then puts divide(@a, @b)
    when '+' then puts @a + @b
    when '-' then puts subtract(@a, @b)
    when '^' then puts exponent(@a, @b)
    end
  end

  def make_neg(a)
    multiply(a, NEG_VAL)
  end

  def subtract(a, b)
    c = make_neg(b)
    a + c
  end

  def multiply(a, b)
    c = 0
    if a < 0 && b > 0
      b.times { c += a }
    elsif b < 0 && a > 0
      a.times { c += b }
    else
      a.abs.times { c += b.abs }
    end
    c
  end

  def exponent(a, b)
    return 'Non-integral answer' if b < 0
    count = 0
    c = 1
    until count == b
      c = multiply(a, c)
      count += 1
    end
    c
  end

  def divide(a, b)
    return 'Not-defined' if b.zero?
    negative = true if a ^ b < 0
    a, b = a.abs, b.abs
    count = 0
    c = 0
    until c >= a
      c += b
      count += 1
    end
    count = make_neg(count) if negative
    c == a ? count : 'Non-integral answer'
  end
end

File: Easy/331_adding_calculator.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int make_neg(int a);
void subtract(int a, int b);
int multiply(int a, int b);
void exponent(int a, int b);
void divide(int a, int b);

int main(void)
{
    int a;
    int b;
    char operator;
    while (scanf("%i %c %i", &a, &operator, &b) == 3) {
        if (operator == '/') {
            divide(a, b);
        }
        if (operator == '*') {
            printf("%i\n", multiply(a, b));
        }
        if (operator == '^') {
            exponent(a, b);
        }
        if (operator == '-') {
            subtract(a, b);
        }
        if (operator == '+') {
            printf("%i\n", a + b);
        }
    }
}

int make_neg(int a)
{
    int min = INT_MIN + 2147483647;
    return multiply(a, min);
}

void subtract(int a, int b)
{
    int c = make_neg(b);
    printf("%i\n", a + c);
}

int multiply(int a, int b)
{
    int c = 0;
    if (a < 0 && b > 0) {
        for (int i = 0; i < b; i++) {
            c += a;
        }
    } else if (b < 0 && a > 0) {
        for (int i = 0; i < a; i++) {
            c += b;
        }
    } else {
        a = abs(a);
        b = abs(b);
        for (int i = 0; i < a; i++) {
            c += b;
        }
    }
    return c;
}

void exponent(int a, int b)
{
    if (b < 0) {
        printf("Non-integral answer\n");
    }
    int count = 0;
    int c = 1;
    while (count < b) {
        c = multiply(a, c);
        count++;
    }
    if (b >= 0) {
        printf("%i\n", c);
    }
}

void divide(int a, int b)
{
    int negative = 1;
    if (b == 0) {
        printf("Not-defined\n");
    }
    if ((a < 0 && b > 0) || (a > 0 && b < 0)) {
        negative = 0;
    }
    a = abs(a);
    b = abs(b);
    int count = 0;
    int c = 0;
    while (c < a) {
        c += b;
        count++;
    }
    if (negative == 0) {
        count = make_neg(count);
    }
    if (c == a && b != 0) {
        printf("%i\n", count);
    } else if (c != a) {
        printf("Non-integral answer\n");
    }
}
